### PR TITLE
Use vectorized T=byte implementations to optimize all MemoryExtensions APIs for T != byte

### PR DESCRIPTION
**Pull request by ahsonkhan**
Related to https://github.com/dotnet/corefx/issues/27487 and partially addresses https://github.com/dotnet/corefx/issues/27379

Builds on top of https://github.com/dotnet/corefx/pull/27859 / https://github.com/dotnet/corefx/pull/28073

**TODO:** ~Add more unit tests and~ measure performance impact

cc @AtsushiKan, @jkotas, @stephentoub, @KrzysztofCwalina

### CODE REVIEW

File: src/Common/src/CoreLib/System/MemoryExtensions.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@
 #if BIT64
 using nuint=System.UInt64;
 #else
-using nuint=System.UInt32;
+using nuint = System.UInt32;
```
Comment (stephentoub) [inline src/Common/src/CoreLib/System/MemoryExtensions.cs]: Nit: this formatting should be consistent with the if

File: src/Common/src/CoreLib/System/MemoryExtensions.cs [hunk 2]
```diff
@@ -186,11 +186,12 @@ public static bool IsWhiteSpace(this ReadOnlySpan<char> span)
         public static int IndexOf<T>(this Span<T> span, T value)
             where T : IEquatable<T>
         {
-            if (typeof(T) == typeof(byte))
+            if (IsTypeNumeric<T>(out int size))
```
Comment (stephentoub) [inline src/Common/src/CoreLib/System/MemoryExtensions.cs]: With `typeof(T) == typeof(byte)`, the JIT will avoid needing to generate any code for the other branch.  Is it able to with this pattern as well?

File: src/Common/src/CoreLib/System/SpanHelpers.Byte.cs [hunk 3]
```diff
@@ -37,13 +37,13 @@ public static int IndexOf(ref byte searchSpace, int searchSpaceLength, ref byte
...
             ref byte valueTail = ref Unsafe.Add(ref value, 1);
-            int valueTailLength = valueLength - 1;
+            nuint valueTailLength = valueLength - 1;
 
             int index = 0;
             for (; ; )
             {
-                Debug.Assert(0 <= index && index <= searchSpaceLength); // Ensures no deceptive underflows in the computation of "remainingSearchSpaceLength".
-                int remainingSearchSpaceLength = searchSpaceLength - index - valueTailLength;
+                Debug.Assert(0 <= index && searchSpaceLength >= index); // Ensures no deceptive underflows in the computation of "remainingSearchSpaceLength".
```
Comment (stephentoub) [inline src/Common/src/CoreLib/System/SpanHelpers.Byte.cs]: Why did you change this?

Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/SpanHelpers.Byte.cs]: I added the `>=(NUInt left, int right)` operator, but not the other way around (i.e. `<=(int left, NUInt right)`) to the NUint wrapper for netfx.

Comment (jkotas) [reply src/Common/src/CoreLib/System/SpanHelpers.Byte.cs]: I think we should add as many operators as necessary to NUInt. so we do not need unnatural workarounds like this.

File: src/Common/src/CoreLib/System/MemoryExtensions.cs [hunk 4]
```diff
@@ -328,11 +295,12 @@ public static void Reverse<T>(this Span<T> span)
         public static int IndexOf<T>(this ReadOnlySpan<T> span, T value)
             where T : IEquatable<T>
         {
-            if (typeof(T) == typeof(byte))
+            if (IsTypeNumeric<T>(out int size))
                 return SpanHelpers.IndexOf(
                     ref Unsafe.As<T, byte>(ref MemoryMarshal.GetReference(span)),
                     Unsafe.As<T, byte>(ref value),
-                    span.Length);
+                    ((nuint)span.Length) * size);
```
Comment (jkotas) [inline src/Common/src/CoreLib/System/MemoryExtensions.cs]: If I am reading this correctly, this is going to return byte index now. Doesn't it need to return the actual item index?

A lot of tests should be failing because of this.

File: src/Common/src/CoreLib/System/SpanHelpers.Byte.cs [hunk 5]
```diff
@@ -494,15 +494,15 @@ public static unsafe int IndexOfAny(ref byte searchSpace, byte value0, byte valu
...
+        public static unsafe int IndexOfAny(ref byte searchSpace, byte value0, byte value1, byte value2, nuint length)
         {
             Debug.Assert(length >= 0);
 
             uint uValue0 = value0; // Use uint for comparisons to avoid unnecessary 8->32 extensions
             uint uValue1 = value1; // Use uint for comparisons to avoid unnecessary 8->32 extensions
             uint uValue2 = value2; // Use uint for comparisons to avoid unnecessary 8->32 extensions
             IntPtr index = (IntPtr)0; // Use UIntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
-            IntPtr nLength = (IntPtr)(uint)length;
+            IntPtr nLength = (IntPtr)length;
```
Comment (jkotas) [inline src/Common/src/CoreLib/System/SpanHelpers.Byte.cs]: We should use nuint instead of IntPtr in the implementation instead of the hacky combination of IntPtrs and pointers.

File: src/Common/src/CoreLib/System/MemoryExtensions.cs (hunk 4 above)
Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: Yes, and they are. Fixing it now.

File: src/Common/src/CoreLib/System/SpanHelpers.Byte.cs (hunk 3 above)
Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/SpanHelpers.Byte.cs]: OK. I will go ahead and add the necessary combinations.

File: src/Common/src/CoreLib/System/SpanHelpers.Byte.cs [hunk 6]
```diff
@@ -981,16 +977,16 @@ private static int LocateFirstFoundByte(Vector<byte> match)
...
             Debug.Assert(firstLength >= 0);
             Debug.Assert(secondLength >= 0);
 
             if (Unsafe.AreSame(ref first, ref second))
                 goto Equal;
 
-            var minLength = firstLength;
-            if (minLength > secondLength) minLength = secondLength;
+            nuint minLength = firstLength;
+            if ((byte*)(IntPtr)minLength > (byte*)(IntPtr)secondLength) minLength = secondLength;
```
Comment (jkotas) [inline src/Common/src/CoreLib/System/SpanHelpers.Byte.cs]: This can be just:

```
if (minLength > secondLength) minLength = secondLength;
```

Comment (jkotas) [reply src/Common/src/CoreLib/System/SpanHelpers.Byte.cs]: (Similar in other places.)

File: src/Common/src/CoreLib/System/MemoryExtensions.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: Yes, the JIT is only generating the necessary code and treating size as a constant.

For byte:
image

For int:
image

File: src/Common/src/CoreLib/System/SpanHelpers.Byte.cs [hunk 7]
```diff
@@ -981,28 +977,28 @@ private static int LocateFirstFoundByte(Vector<byte> match)
...
+        public static unsafe int SequenceCompareTo(ref byte first, nuint firstLength, ref byte second, nuint secondLength)
         {
             Debug.Assert(firstLength >= 0);
             Debug.Assert(secondLength >= 0);
 
             if (Unsafe.AreSame(ref first, ref second))
                 goto Equal;
 
-            var minLength = firstLength;
+            nuint minLength = firstLength;
```
Comment (jkotas) [inline src/Common/src/CoreLib/System/SpanHelpers.Byte.cs]: SequenceEqual has almost identical code, can use the same cleanup.

File: src/Common/src/CoreLib/System/MemoryExtensions.cs (hunk 2 above)
Comment (jkotas) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: What does it look before/after for `Span<string>` ?

Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: It is bad for `Span<string>`. Before, all checks are avoided. Now, the IsTypeNumeric method is called with all the checks.

image

**Comment by AndyAyersMS** [comment]
I think if you add in the `default(T) == null` check for ref types as an explicit case it should work.

File: src/Common/src/CoreLib/System/MemoryExtensions.cs [hunk 8]
```diff
@@ -1304,5 +1305,36 @@ public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> secon
                 value, comparer);
             return BinarySearch(span, comparable);
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsTypeNumeric<T>(out int size)
```
Comment (AtsushiKan) [inline src/Common/src/CoreLib/System/MemoryExtensions.cs]: A better name would be something like `IsComparableAsBytes`. `IsTypeNumeric` sounds like it should include types like R4 and R8 and Decimal (which it shouldn't - you can't use byte compare to compare those types) Also, `char` isn't really "numeric."

You could also test for IntPtr/UIntPtr here.

File: src/Common/src/CoreLib/System/MemoryExtensions.cs [hunk 9]
```diff
@@ -1304,5 +1305,36 @@ public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> secon
...
+                return true;
+            }
+
+            if (typeof(T) == typeof(long) || typeof(T) == typeof(ulong))
+            {
+                size = sizeof(long);
+                return true;
+            }
+
+            size = 0;
```
Comment (AtsushiKan) [inline src/Common/src/CoreLib/System/MemoryExtensions.cs]: Nit: prefer `size = default;` here as the intent isn't that the size is zero, the intent is that the size is uninteresting.

File: src/System.Memory/src/System/NUint.cs [hunk 10]
```diff
@@ -29,6 +29,68 @@ internal struct NUInt
...
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static NUInt operator -(NUInt left, int right)
+        {
+            unsafe { return (sizeof(IntPtr) == 4) ? new NUInt(((uint)left._value) - (uint)right) : new NUInt(((ulong)left._value) - (uint)right); }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static NUInt operator *(NUInt left, NUInt right)
+        {
+            unsafe { return (sizeof(IntPtr) == 4) ? new NUInt(((uint)left._value) * (uint)right._value) : new NUInt(((ulong)left._value) * (uint)right._value); }
```
Comment (AtsushiKan) [inline src/System.Memory/src/System/NUint.cs]: The rightmost cast of `right` to `NUint` will cause value loss. Same with other operators that take a `NUint right`

File: src/Common/src/CoreLib/System/MemoryExtensions.cs (hunk 8 above)
Comment (AtsushiKan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: Might be better just to have to return a `nuint` size. You're casting it to `nuint` everywhere you use it anyway.

File: src/Common/src/CoreLib/System/MemoryExtensions.cs [hunk 11]
```diff
@@ -292,31 +276,14 @@ public static bool SequenceEqual<T>(this Span<T> first, ReadOnlySpan<T> second)
         public static int SequenceCompareTo<T>(this Span<T> first, ReadOnlySpan<T> second)
             where T : IComparable<T>
         {
-            if (typeof(T) == typeof(byte))
```
Comment (AtsushiKan) [inline src/Common/src/CoreLib/System/MemoryExtensions.cs]: I don't think you can validly apply this optimization to `SequentialCompare` - comparing unsigned bytes one at a time isn't the same as comparing elements using the proper Compare algorithm.

Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: Good point. I will revert this. We need to add more tests for T != byte for SequenceCompareTo.

Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: https://github.com/dotnet/corefx/issues/28118

File: src/Common/src/CoreLib/System/MemoryExtensions.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: > I think if you add in the `default(T) == null` check for ref types as an explicit case it should work.

Andy's suggestion worked well:

image

Comment (jkotas) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: It is definitely better, but there is still extra left over cruft. Should we rather look into generating the code using template, like we do for `Vector<T>` ?

Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: > Should we rather look into generating the code using template

Seems excessive for this especially since the code is identical for most of the types. Alternatively, we could just remove the IsTypeComparableAsBytes method and duplicate the code within the 6 call sites. Would a few extra instructions matter here (specifically for reference types like string)? There isn't any extra overhead for types like byte/int.

Comment (jkotas) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: https://github.com/dotnet/corefx/pull/28080#discussion_r174667868 showed extra instructions even for non-reference case.

Also the real code may suffer from this more than microbechmarks. The JIT has a fixed limits on the amount of code that it optimizes or number of local variables that it optimizes. All the extra complexity counts against these limits.

I would also make the AggressiveInlining code as streamlined as possible. In this case, it means duplicating the code withing the 6 call sites.

Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: >  showed extra instructions even for non-reference case.

The extra instructions in the byte case were a copy/paste mistake. Just like int, the byte case doesn't have overhead:
image

> Also the real code may suffer from this more than microbechmarks. The JIT has a fixed limits on the amount of code that it optimizes or number of local variables that it optimizes. All the extra complexity counts against these limits.
I would always make the AggressiveInlining code as streamlined as possible. In this case, I think it means duplicating the code within the 6 call sites.

OK, I will duplicate the code then.

**Comment by ahsonkhan** [comment]
Sample performance impact (chose StartsWith):
image

**Comment by jkotas** [comment]
The regressions for small value sizes are not good. I would expect that StartsWith will be typically used with relatively small sizes.

**Comment by ahsonkhan** [comment]
> The regressions for small value sizes are not good. I would expect that StartsWith will be typically used with relatively small sizes.

The regression is only there for length == 1. I want to run some tests and collect data, and if that case is very common, we can consider special casing it.

image

**Comment by jkotas** [comment]
Ok, this looks better.

File: src/Common/src/CoreLib/System/MemoryExtensions.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: > The JIT has a fixed limits on the amount of code that it optimizes or number of local variables that it optimizes. All the extra complexity counts against these limits. I would always make the AggressiveInlining code as streamlined as possible. In this case, I think it means duplicating the code within the 6 call sites.

> OK, I will duplicate the code then.

On second thought, I am not sure if this much duplication is worth it, given it would only add a few instructions specific to reference types and the code becomes long and cumbersome. If the JIT, in some cases, is unable to optimize this, then how will duplicating the code within the method help avoid that? I am not sure how we are reducing the complexity here. @jkotas, thoughts? Are there other benefits than having one less method that is marked with AggressiveInlining?

Comment (jkotas) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: If you inline the code manually, the JIT turns a lot of things into constants right away. The dead code can be pruned quickly, and the trees created are generally simple. If you let the JIT to do the inlining, there are complex trees created first and that the JIT needs to work hard to simplify. It has likely negative impact on both throughput; and on the code quality. I think you should be able to see the impact on code quality if you just manually unroll the microbenchmark to calls these methods say 50x. This should be enough to hit the JITs thresholds on too complex code. Measure the performance with both with and without manual inlining.

@AndyAyersMS Do you have an opinion whether it is better to inline manually or whether it is better to have the nested aggressively inlined methods here?

Comment (ahsonkhan) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: > I think you should be able to see the impact on code quality if you manually unroll the microbenchmark to calls these methods say 50x. This should be enough to hit the JITs thresholds on too complex code. Measure the performance with both with and without manual inlining.

I call the method ~50x times (`result &= StartsWith(...)`) within each iteration of the benchmark. I wasn't able to see a performance difference:
image

image

**Comment by ahsonkhan** [comment]
OSX x64 Debug Build
https://mc.dot.net/#/user/ahsonkhan/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/be5aaa6aafd5a19086974af807cb66588586e825/workItem/System.Reflection.Metadata.Tests/wilogs
OSX.1012.Amd64.Open:Debug-x64
https://github.com/dotnet/corefx/issues/27375

@dotnet-bot test OSX x64 Debug Build

File: src/Common/src/CoreLib/System/MemoryExtensions.cs (hunk 2 above)
Comment (AndyAyersMS) [reply src/Common/src/CoreLib/System/MemoryExtensions.cs]: The jit has gotten better at early (importer) pruning of dead code so often times we only import a slice of a method. We're also more aggressive about forwarding values into inline bodies and out of returns. 

So I think the nested aggressive inline is ok here from the jit's standpoint, and it probably makes the code more readable/maintainable.

**Comment by ahsonkhan** [comment]
@dotnet-bot test OSX x64 Debug Build

**Comment by ahsonkhan** [comment]
cc @Anipik, @safern - I was expecting a mirror PR in coreclr. Do you know what's blocking it?

**Comment by Anipik** [comment]
@ahsonkhan  Mirror was blocked but now its up again. mirror has already opened PRs. This will be picked after the opened PRs has been merged
